src/cms/models/pages/abstract_base_page.py

```diff
@@ -0,0 +1,208 @@
+        and filters them to the requested :class:`~cms.models.languages.language.Language` code.
+
+        :param language_code: The code of the desired :class:`~cms.models.languages.language.Language`
+        :type language_code: str
+
+        :return: The page translation in the requested :class:`~cms.models.languages.language.Language` or :obj:`None`
+                 if no translation exists
+        :rtype: ~cms.models.pages.page_translation.PageTranslation
+        """
+        return self.translations.filter(language__code=language_code).first()
+
+    def get_first_translation(self, priority_language_codes=None):
+        """
+        Helper function for page labels, second level paths etc. where the ancestor translation might not exist
+        This function uses the reverse foreign key ``self.translations`` to get all translations of ``self``
+        and filters them to the first requested :class:`~cms.models.languages.language.Language` code that matches.
+        So a lower list index means a higher priority.
+
+        :param priority_language_codes: A list of :class:`~cms.models.languages.language.Language` codes,
+                                        defaults to ``None``
+        :type priority_language_codes: list [ str ], optional
+
+        :return: The first page translation which matches one of the :class:`~cms.models.languages.language.Language`
+                 given or :obj:`None` if no translation exists
+        :rtype: ~cms.models.pages.page_translation.PageTranslation
+        """
+        # Taking [] directly as default parameter would be dangerous because it is mutable
+        if not priority_language_codes:
+            priority_language_codes = []
+        for language_code in priority_language_codes + ["en-us", "de-de"]:
+            if self.translations.filter(language__code=language_code).exists():
+                return self.translations.filter(language__code=language_code).first()
+        return self.translations.first()
+
+    def get_public_translation(self, language_code):
+        """
+        This function retrieves the newest public translation of a page.
+
+        :param language_code: The code of the requested :class:`~cms.models.languages.language.Language`
+        :type language_code: str
+
+        :return: The public translation of a page
+        :rtype: ~cms.models.pages.page_translation.PageTranslation
+        """
+        return self.translations.filter(
+            language__code=language_code,
+            status=status.PUBLIC,
+        ).first()
+
+    @abstractmethod
+    def get_absolute_url(self):
+        """
+        This helper function returns the absolute url to the editing form of a page
+
+        :return: The absolute url of a page form
+        :rtype: str
+        """
+        pass
+
+    # TODO should this be implemented, passed or not be present here at all?
```

Comment:
Should this be implemented, passed or not be present here at all? Not sure if it is necessary for imprints at all (since there is probably not more than one anyway). Same for `get_archived` below.